### ISSUE TITLE
Ugly frame rate with scrcpy and onn 4k 2003 box

### ISSUE THREAD

**Issue by nc88keyz** [Genymobile/scrcpy#5354]
Here is the issue: frame rate is terribly slow over adb, Ive tried all the flags and the best can be seen is around 9-10 frames refresh rate. I have tried direct, over wireguard, over tail scale. does not matter. Here is what is really happening though.  After much trial and error. Anydesk produces a near real time remote layer with very acceptable frame rate. The only work around I can do for maintenance on these installation is the following. Split screen with scrcpy on the right, and anydesk pinned on the right. With the mouse hovering over the scrcpy window. The visual speed is real time with all mouse movements and keyboard input.  My hope ?

Can this be fixed in scrcpy? or is something broken on adb protocol for the Onn 2023 4k box.  I realize its a $20 value box in the USA, but the wifi streams fine , its relatively snappy so i don't see what is broken with scrcpy.  Maybe something can be tweaked for it. On the same network a tivostream and multiple shields do not have this issue. 

Specs on the Onn 2023 4k stick, [https://www.aftvnews.com/comparison-of-walmarts-onn-hd-stick-onn-4k-box-and-onn-4k-pro-streaming-devices-with-google-tv/](url)

**Comment by rom1v** [comment]
What command do you execute? Is it over USB? Over wifi?

Which resolution? Which bitrate?

What is the result with `--print-fps`?

If you record and play the file (`scrcpy --record=file.mp4 …`), is it laggy?

**Comment by nc88keyz** [comment]
ill get those for you, but i didnt see different in bitrates that I tried, not sure what I am looking for on --print-fps, and no did not try record=file.mp4, but can on a LAN. as it did the same as via remote basically with my personal onn 2023 4k for testing. 
I tried both usb and wifi.

**Comment by nc88keyz** [comment]
ok the following works well, at about 1 to 1.5 fps
-b2M --max-size 720
not sure about fps now? will update

**Comment by rom1v** [comment]
Also try another encoder: https://github.com/Genymobile/scrcpy/blob/master/doc/video.md#encoder

> 7-8fps

What is on the device screen? (scrcpy never captures a frame if nothing has changed, so for example on a static screen you get 0fps). Is it smooth on the device itself?

> or is something broken on adb protocol

All of this has nothing to do with adb, it's the capture+encoding on the device by MediaCodec (and the encoder implementation).

**Comment by nc88keyz** [comment]
It was just projectivity launcher, and moving arrows back and forth up down, ocassioanlly clicking enter on the cog. , 0,1,0,0 etc. with 1080, at 720  and 2M it gets around 7,8 for same activity on launcher. 

i tried the encoders thing before this is what is returned with list 

[server] INFO: List of video encoders:
    --video-codec=h264 --video-encoder='c2.android.avc.encoder'
    --video-codec=h264 --video-encoder='OMX.google.h264.encoder'
[server] INFO: List of audio encoders:
    --audio-codec=opus --audio-encoder='c2.android.opus.encoder'
    --audio-codec=aac --audio-encoder='c2.android.aac.encoder'
    --audio-codec=aac --audio-encoder='OMX.google.aac.encoder'

**Comment by rom1v** [comment]
Did you try both video encoders?

**Comment by nc88keyz** [comment]
INFO: 1 fps
INFO: 1 fps
INFO: 1 fps
INFO: 1 fps
INFO: 1 fps
INFO: 1 fps
INFO: 1 fps
INFO: 1 fps
INFO: 1 fps
INFO: 1 fps
INFO: 1 fps
INFO: 1 fps
INFO: 1 fps
INFO: 1 fps
INFO: 1 fps
INFO: 1 fps
INFO: 1 fps
INFO: 2 fps
INFO: 1 fps
INFO: 1 fps
INFO: 1 fps
INFO: 2 fps
INFO: 1 fps
INFO: 2 fps
INFO: 1 fps
INFO: 1 fps
INFO: 2 fps
INFO: 1 fps
INFO: 1 fps

at 2M 1080 this is the output  the audio does appear to be almost real time, with acceptable lag

**Comment by rom1v** [comment]
With `--video-encoder=c2.android.avc.encoder` or `--video-encoder=OMX.google.h264.encoder`?

**Comment by rom1v** [comment]
Also, out of curiosity, does `--turn-screen-off` change anything?

**Comment by nc88keyz** [comment]
encoders do not make a difference at 720p , 7-8 fps output. will update with turn screen off, i think it didnt work before but could be wrong

**Comment by rom1v** [comment]
7-8 fps is clearly not normal. It should be 60fps.

Here, this is a capture/encoding issue (if you record, you'll get low frame rate too in the recorded file), but I don't know why.

> when anydesk output is near admirable.

Anydesk running on the very same device? Every app capturing the screen should also use MediaCodec, so if it behaves differently it probably gives different options.

You can play with options here:
https://github.com/Genymobile/scrcpy/blob/665ccb32f5306ebd866dc0d99f4d08ed2aeb91c3/server/src/main/java/com/genymobile/scrcpy/video/SurfaceEncoder.java#L230-L246
(you can change values with `--video-codec-options`)

Does it work better with `--no-audio`?

**Comment by nc88keyz** [comment]
I agree its not normal, and every other android tv devices in my house does not exibit these issues. have not seen anything about 7,8 fps unless you drop to 480, 1M which is unbearable. I will say this the connection is around 48-56Mbps over wifi on an asus Xt8 mesh point. They go up to 6E if the device supports it. I think i have two devices , that do, not that that should matter, but it was the same over usb as well, with no noticable. Someone has to have an Onn 2023 4k walmart special that can test out as well. Ive gotten the same results on two different LANS and over WAN as well. via native RDP, rustdesk, anydesk, via remote, and even remote desktop to usb adb , ( had my dad order a usb2go adapter and we ended up just powering it off the laptop itself and doing without the usb2go adapter as the usb a to a cable we ordered would not recognize on the laptop. Also tried on my own network, I bought one just to test out and never returned it to Walmart, to help out a friend and my dad as well. They all work great and snappy with exception to the encoding through scrcpy and adb. I dont know what else to troubleshoot. I dont guess you can get your hands on one of these boxes?

**Comment by nc88keyz** [comment]
> 7-8 fps is clearly not normal. It should be 60fps.
> 
> Here, this is a capture/encoding issue (if you record, you'll get low frame rate too in the recorded file), but I don't know why.
> 
> > when anydesk output is near admirable.
> 
> Anydesk running on the very same device? Every app capturing the screen should also use MediaCodec, so if it behaves differently it probably gives different options.
> 
> You can play with options here:
> 
> https://github.com/Genymobile/scrcpy/blob/665ccb32f5306ebd866dc0d99f4d08ed2aeb91c3/server/src/main/java/com/genymobile/scrcpy/video/SurfaceEncoder.java#L230-L246
> 
> 
> (you can change values with `--video-codec-options`)
> Does it work better with `--no-audio`?

no--difference-with--no-audio-lol

**Comment by rom1v** [comment]
@nc88keyz Too much information. The problem is a capture/encoding issue, so we don't care about network, bandwith, usb, routers…

Do you get smooth mirroring (60fps) on the very same box with another software?

It seems your device does not have any hardware encoder, that's why it's slow: it encodes in software. But if it works with another software on the very same device, then I'm wrong.

**Comment by nc88keyz** [comment]
by device you mean the source android device, yes? 

Is that even possible for the Onn box to not have an encoder that functions?  

so, anydesk is an android app I. the playstore. It's installed on the device ( onn) and the viewer is x64 anydesk client on Win11.  It's can stream very fluidly in comparison to the encoder output recieved by scrcpy. I think I mentioned before that the capture is better but not 60fps by any means.  Is there a way to verify no encoder is present or disabled?

**Comment by shubhamsinghshubham777** [comment]
I'm facing the same issue. Very laggy experience on version `2.7`. I've tried `-m 768` (even lower values) but even after having good enough bandwidth (200 mbps), I can't seem to get a smooth screen mirroring experience.

**Comment by pythcon** [comment]
@nc88keyz , did you ever figure out a solution to this? Exhibiting the same behavior on my Onn 4K Pro boxes.
